### PR TITLE
Add support for VisiView STK files

### PR DESCRIPTION
**Pull request by dgault**
This PR is in response to forum thread https://forum.image.sc/t/issue-with-stk-timelapse-import-to-omero-or-fiji-with-bio-formats/54644

The VisiView software is generating STK files (no associated ND file) which can be handled by the existing MetaMorphReader, this PR simply allows VisiView as valid software for this readers `isThisType` method. Without this PR they will be read by the TiffDelegateReader and only a single plane of a time series will be displayed.

A follow up config PR will be opened once I have completed transfer of the sample file.

### CODE REVIEW

File: components/formats-gpl/src/loci/formats/in/MetamorphReader.java [hunk 1]
```diff
@@ -235,7 +235,8 @@ public boolean isThisType(RandomAccessInputStream stream) throws IOException {
     if (ifd == null) return false;
     String software = ifd.getIFDTextValue(IFD.SOFTWARE);
     boolean validSoftware = software != null &&
-      software.trim().toLowerCase().startsWith("metamorph");
+      (software.trim().toLowerCase().startsWith("metamorph") || 
+          software.trim().toLowerCase().startsWith("visiview"));
```
Comment (sbesson) [inline components/formats-gpl/src/loci/formats/in/MetamorphReader.java:239]: Should this list be extracted into `public static` constant of the reader?

File: components/formats-gpl/src/loci/formats/in/MetamorphReader.java [hunk 2]
```diff
@@ -105,6 +105,8 @@
   public static final String[] ND_SUFFIX = {"nd", "scan"};
   public static final String[] STK_SUFFIX = {"stk", "tif", "tiff"};
 
+  public static final String[] validSoftware = {"metamorph", "visiview"};
```
Comment (sbesson) [inline components/formats-gpl/src/loci/formats/in/MetamorphReader.java:108]:
```suggestion
  public static final String[] VALID_SOFTWARE = {"metamorph", "visiview"};
```

File: components/formats-gpl/src/loci/formats/in/MetamorphReader.java [hunk 3]
```diff
@@ -2303,4 +2302,12 @@ else if (version.equals(NDINFOFILE_VER2)) {
     return formatSuffix;
   }
 
+  private static boolean isValidSoftware(String software) {
+    if (software == null) return false;
+    for (int i = 0; i < validSoftware.length; i++) {
```
Comment (sbesson) [inline components/formats-gpl/src/loci/formats/in/MetamorphReader.java:2307]:
```suggestion
    for (int i = 0; i < VALID_SOFTWARE.length; i++) {
```

File: components/formats-gpl/src/loci/formats/in/MetamorphReader.java [hunk 4]
```diff
@@ -2303,4 +2302,12 @@ else if (version.equals(NDINFOFILE_VER2)) {
...
+      if (software.trim().toLowerCase().startsWith(validSoftware[i])) return true;
```
Comment (sbesson) [inline components/formats-gpl/src/loci/formats/in/MetamorphReader.java:2308]:
```suggestion
      if (software.trim().toLowerCase().startsWith(VALID_SOFTWARE[i])) return true;
```

**Review by sbesson — COMMENTED** [review]
A few suggestion to unify the constant naming styles with existing behavior. Otherwise, looks good and I'll retest tomorrow after the commit is validated by the CI builds

**Comment by dgault** [comment]
Seems to be failures from the repo tests with the latest commits: https://merge-ci.openmicroscopy.org/jenkins/job/BIOFORMATS-test-folder/lastFailedBuild/console

**Comment by dgault** [comment]
Apparently we did have some VisiView files in the repo already. I will have to update the configs

**Comment by dgault** [comment]
The failures with this PR are related to a single TIFF file (https://merge-ci.openmicroscopy.org/jenkins/job/BIOFORMATS-test-folder/lastFailedBuild/console).

The file had been read as a single plane TIFF with 1C and 1Z. 

With this PR included, and run with Java 8, the file is now associated with additional files due to matching prefix. This is resulting in the additional files being read as a 2 channel image.
```
filename = /Volumes/Seagate Desktop Drive/SampleImages/metamorph/jan-eglinger/benchmark_v1_2018_x64y64z1c1s1t1.tif
Used files:
	/Volumes/Seagate Desktop Drive/SampleImages/metamorph/jan-eglinger/benchmark_v1_2018_x64y64z1c2s1t11.nd
	/Volumes/Seagate Desktop Drive/SampleImages/metamorph/jan-eglinger/benchmark_v1_2018_x64y64z1c2s1t11_w1Laser4054BD4BP.TIF
	/Volumes/Seagate Desktop Drive/SampleImages/metamorph/jan-eglinger/benchmark_v1_2018_x64y64z1c2s1t11_w2Laser4914BD4BP.TIF
Series count = 1
Series #0 :
	Image count = 2
	RGB = false (1) 
	Interleaved = false
	Indexed = false (false color)
	Width = 64
	Height = 64
	SizeZ = 1
	SizeT = 1
	SizeC = 2
	Tile size = 64 x 64
	Thumbnail size = 64 x 64
	Endianness = intel (little)
	Dimension order = XYZCT (certain)
	Pixel type = uint16
	Valid bits per pixel = 16
	Metadata complete = true
	Thumbnail series = false
```

More unusually when run with a later Java version of then the file is now associated with a different dataset with a sizeC of 2 and sizeZ of 5:
```
filename = /Users/dgault/Documents/Sample Images/metamorph/jan-eglinger/benchmark_v1_2018_x64y64z1c1s1t1.tif
Used files:
	/Users/dgault/Documents/Sample Images/metamorph/jan-eglinger/benchmark_v1_2018_x64y64z1c2s4t12.nd
	/Users/dgault/Documents/Sample Images/metamorph/jan-eglinger/benchmark_v1_2018_x64y64z1c2s4t12_w1Laser4054BD4BP.STK
	/Users/dgault/Documents/Sample Images/metamorph/jan-eglinger/benchmark_v1_2018_x64y64z1c2s4t12_w2Laser4914BD4BP.STK
Series count = 1
Series #0 :
	Image count = 10
	RGB = false (1) 
	Interleaved = false
	Indexed = false (false color)
	Width = 64
	Height = 64
	SizeZ = 5
	SizeT = 1
	SizeC = 2
	Tile size = 64 x 64
	Thumbnail size = 64 x 64
	Endianness = intel (little)
	Dimension order = XYZCT (certain)
	Pixel type = uint16
	Valid bits per pixel = 16
	Metadata complete = true
	Thumbnail series = false
```

When attempting to open the file you are then presented with a FormatException:
```
Exception in thread "main" loci.formats.FormatException: Invalid image number: 4 (series=0, imageCount=4)
	at loci.formats.FormatTools.checkPlaneNumber(FormatTools.java:1013)
	at loci.formats.FormatTools.checkPlaneParameters(FormatTools.java:1001)
	at loci.formats.in.MinimalTiffReader.openBytes(MinimalTiffReader.java:289)
	at loci.formats.in.MetamorphReader.openBytes(MetamorphReader.java:296)
	at loci.formats.in.MetamorphReader.openBytes(MetamorphReader.java:325)
	at loci.formats.FormatReader.openBytes(FormatReader.java:919)
	at loci.formats.ImageReader.openBytes(ImageReader.java:449)
```

Pushing this PR as it will require reworking of the file grouping to ensure that these datasets are not incorrectly being grouped.

**Comment by dgault** [comment]
Closing this PR as we have not seen any further requests for these particular files since the original request.

**Pull request closed without merging**